### PR TITLE
fix(MainchainGatewayV3): more checks on invalid signer and null total weight

### PR DESCRIPTION
**Pull request by TuDo1403**
### Description
This PR enhances `MainchainGatewayV3::submitWithdrawal` function by:
- Revert when `totalVoteWeight` is `0`.
- Revert when `minimumVoteWeight` is `0`.
- Revert when `ECDSA` v, r, s are non-unique (malleable).
- Revert when signer weight is `0`.

### Checklist
- [x] I have clearly commented on all the main functions following the [NatSpec Format](https://docs.soliditylang.org/en/v0.8.0/natspec-format.html)
- [x] The box that allows repo maintainers to update this PR is checked
- [x] I tested locally to make sure this feature/fix works

### CODE REVIEW

File: src/mainchain/MainchainGatewayV3.sol [hunk 1]
```diff
@@ -309,16 +310,19 @@ contract MainchainGatewayV3 is
       address signer;
       address lastSigner;
       Signature memory sig;
-      uint256 weight;
+      uint256 accWeight;
```
Comment (nxqbao) [inline src/mainchain/MainchainGatewayV3.sol]: https://www.abbreviations.com/abbreviation/accumulative

```suggestion
      uint256 accumWeight;
```

File: src/mainchain/MainchainGatewayV3.sol [hunk 2]
```diff
@@ -461,8 +465,9 @@ contract MainchainGatewayV3 is
   /**
    * @inheritdoc GatewayV3
    */
-  function _getTotalWeight() internal view override returns (uint256) {
-    return _totalOperatorWeight;
+  function _getTotalWeight() internal view override returns (uint256 totalWeight) {
+    totalWeight = _totalOperatorWeight;
+    if (totalWeight == 0) revert ErrNullTotalWeightProvided(msg.sig);
```
Comment (nxqbao) [inline src/mainchain/MainchainGatewayV3.sol:470]: should we check minimumVoteWeight as well?

File: src/mainchain/MainchainGatewayV3.sol [hunk 3]
```diff
@@ -309,16 +310,19 @@ contract MainchainGatewayV3 is
...
-        signer = ecrecover(receiptDigest, sig.v, sig.r, sig.s);
+        signer = ECDSA.recover({ hash: receiptDigest, v: sig.v, r: sig.r, s: sig.s });
         if (lastSigner >= signer) revert ErrInvalidOrder(msg.sig);
 
         lastSigner = signer;
 
-        weight += _getWeight(signer);
-        if (weight >= minimumWeight) {
+        uint256 w = _getWeight(signer);
+        if (w == 0) revert ErrInvalidSignature(signer, w, sig);
```
Comment (huyhuynh3103) [inline src/mainchain/MainchainGatewayV3.sol]:
```suggestion
        if (w == 0) revert ErrInvalidSigner(signer, w, sig);
```

File: src/mainchain/MainchainGatewayV3.sol (hunk 1 above)
Comment (TuDo1403) [reply src/mainchain/MainchainGatewayV3.sol]: fixed.

File: src/mainchain/MainchainGatewayV3.sol (hunk 3 above)
Comment (TuDo1403) [reply src/mainchain/MainchainGatewayV3.sol]: fixed.

File: src/mainchain/MainchainGatewayV3.sol (hunk 2 above)
Comment (TuDo1403) [reply src/mainchain/MainchainGatewayV3.sol:470]: updated.

**Review by nxqbao — APPROVED** [review]

**Review by khanhtaskymaviscom — APPROVED** [review]

**Pull request merged**